### DISCUSSION TITLE
SML does not work: sml memory error!

### DISCUSSION THREAD

**Discussion by juergiboy** [arendst/Tasmota#17855] [Q&A] [answered]
Hi guys,
you helped me a lot with my problem yesterday, today I unfortunately have a new one.

I build a tasmota for SML with these defines:
#ifdef USE_TLS
#undef USE_TLS
#endif
#define USE_TLS
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
#ifndef SML_REPLACE_VARS
#define SML_REPLACE_VARS
#endif
#ifndef USE_SML_SCRIPT_CMD
#define USE_SML_SCRIPT_CMD
#endif
#ifndef SML_MAX_VARS
#define SML_MAX_VARS 20
#endif

and the board esp8266_4M2M

I added a minimal script:
>D
>B
=>sensor53 r
>M 1
+1,3,v,0,9600,Test

As a result i get this:
11:41:03.117 Script: nv=0, tv=0, vns=0, vmem=16, smem=8192
11:41:03.119 Script: performs "sensor53 r"
11:41:03.126 SNS: Hardware Serial
11:41:03.127 sml memory error!
11:41:03.129 RSL: RESULT = {"Time":"2023-02-02T11:41:03","SML":{"CMD":"restart"}}

I have tried with two different devices - the result is the same.
Can anyone tell what's my problem? Thanks

**Comment by gemu2015** [comment]
1. if you use the latest dev, you do not need some of these defines, most of them are default now
SML_REPLACE_VARS  = default now
USE_SML_SCRIPT_CMD  = default now
USE_TLS is also default from todays dev version 

SML_MAX_VARS is automatically determinated, no need to define

please post your script

did you forgot the # char at the end of the script ?

**Reply by juergiboy** [reply]
Thanks for your hints, i will adjust.
Here my script again, it is not quite complete above:

>D
>B
=>sensor53 r
>M 1
+1,3,v,0,9600,Test
#

**Reply by juergiboy** [reply]
ok, the '>' and '#' are swallowed by the system

**Reply by gemu2015** [reply]
you need the termination char # at the end of the descriptor

```
>D
>B
=>sensor53 r
>M 1
+1,3,v,0,9600,Test
#
```

**Reply by juergiboy** [reply]
`>D`
`>B`
`=>sensor53 r`
`>M 1`
`+1,3,v,0,9600,Test`
`#`

**Reply by juergiboy** [reply]
yes, i did - the result ist the same

**Reply by gemu2015** [reply]
ah i see the problem now.  since not a single decoder line is added the memory check fails.

add at least one decoder line!

**Reply by juergiboy** [reply]
yes, that's it!
Thanks

**Comment by juergiboy** [comment, marked as answer]
Because there ist a question, i put in here the detailed solution:

In the M section of the script you define your meter like this
```
>M 1
+1,3,s,16,9600,Test
```

to prevent the memory problem you have to define at least one decoder line like this
`
1,77078181c78203ff@#,Service ID,,Meter_id,0
`

for details look here: [https://tasmota.github.io/docs/Smart-Meter-Interface/](url)
